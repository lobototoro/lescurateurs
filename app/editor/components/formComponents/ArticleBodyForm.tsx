export function ArticleBodyForm() {
  return (
    <>
      <label htmlFor="title">Title</label>
      <input type="text" id="title" name="title" />
      <label htmlFor="body">Body</label>
      <input type="text" id="body" name="body" />
    </>
  );
}