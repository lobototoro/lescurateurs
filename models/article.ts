export enum UrlsTypes {
  WEBSITE = "website",
  VIDEOS = "videos",
  AUDIO = "audio",
  SOCIAL = "social",
  IMAGE = "image",
}

export interface Article {
  title: string;
  introduction: string | null;
  main: string;
  mainAudioUrl: string | null;
  urlToMainIllustration: string;
  publishedAt: string | null;
  createAt: string;
  updateAt: string;
  author: string;
  author_email: string;
  urls: {
    type: UrlsTypes;
    url: string;
  }[];
}