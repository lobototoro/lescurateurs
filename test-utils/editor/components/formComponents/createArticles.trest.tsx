import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import CreateArticleForm from "@/app/editor/components/formComponents/createArticles";
import { createArticleAction } from "@/app/articleActions";
import { urlsToArrayUtil } from "@/lib/utility-functions";

vi.mock("@/app/articleActions", () => ({
  createArticleAction: vi.fn(),
}));

vi.mock("@/lib/utility-functions", () => ({
  urlsToArrayUtil: vi.fn(() => []),
}));

describe("CreateArticleForm", () => {
  it("renders the form with all required fields", () => {
    render(<CreateArticleForm />);

    expect(screen.getByLabelText(/title/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/introduction/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/main/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/main audio url/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/url to main illustration/i)).toBeInTheDocument();
  });

  it("validates form fields and shows errors", async () => {
    render(<CreateArticleForm />);

    fireEvent.click(screen.getByRole("button", { name: /submit/i }));

    await waitFor(() => {
      expect(screen.getByText(/title is required/i)).toBeInTheDocument();
      expect(screen.getByText(/introduction is required/i)).toBeInTheDocument();
    });
  });

  it("submits the form with valid data", async () => {
    render(<CreateArticleForm />);

    fireEvent.change(screen.getByLabelText(/title/i), { target: { value: "Test Title" } });
    fireEvent.change(screen.getByLabelText(/introduction/i), { target: { value: "Test Introduction" } });
    fireEvent.change(screen.getByLabelText(/main/i), { target: { value: "Test Main Content" } });
    fireEvent.change(screen.getByLabelText(/main audio url/i), { target: { value: "http://example.com/audio.mp3" } });
    fireEvent.change(screen.getByLabelText(/url to main illustration/i), { target: { value: "http://example.com/image.jpg" } });

    fireEvent.click(screen.getByRole("button", { name: /submit/i }));

    await waitFor(() => {
      expect(createArticleAction).toHaveBeenCalled();
    });
  });

  it("adds and removes URL inputs", () => {
    render(<CreateArticleForm />);

    const addButton = screen.getByRole("button", { name: /add url/i });
    const removeButton = screen.getByRole("button", { name: /remove url/i });

    fireEvent.click(addButton);
    expect(urlsToArrayUtil).toHaveBeenCalled();

    fireEvent.click(removeButton);
    expect(urlsToArrayUtil).toHaveBeenCalled();
  });

  it("opens and closes the modal on form submission", async () => {
    render(<CreateArticleForm />);

    const modal = screen.getByTestId("form-sent-modal");
    expect(modal).not.toHaveClass("is-active");

    fireEvent.click(screen.getByRole("button", { name: /submit/i }));

    await waitFor(() => {
      expect(modal).toHaveClass("is-active");
    });

    fireEvent.click(screen.getByRole("button", { name: /close/i }));
    expect(modal).not.toHaveClass("is-active");
  });
});