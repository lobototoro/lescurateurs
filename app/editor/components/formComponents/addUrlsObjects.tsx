"use client";

// import { useState } from "react";
import { UrlsTypes } from "@/models/article";
import { UrlObjectIterator } from "./urlObjectIter";

export function AddUrlsObjects({
  urls,
  setUrls,
  addInputs,
  removeInputs
}: {
  urls: { type: UrlsTypes; url: string; credits?: string }[];
  setUrls: React.Dispatch<React.SetStateAction<any>>;
  addInputs: () => void;
  removeInputs: () => void;
}) {

  return (
    <>
    <div className="field">
      {urls?.map(({ type, url, credits }, index: number) => (
        
          <UrlObjectIterator
            key={`add-url-${index}`}
            type={type || UrlsTypes.WEBSITE}
            url={url || ''}
            credits={credits || ''}
            addUrls={setUrls}
          />
          
      ))}
      <div className="button is-primary" onClick={addInputs}>
            +
      </div>
      <div className="button is-primary" onClick={removeInputs}>
            -
      </div>
      </div>
    </>
  );
};