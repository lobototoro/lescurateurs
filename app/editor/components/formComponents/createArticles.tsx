'use client';
import {
  useActionState,
  useEffect,
  startTransition,
  useRef,
  JSX,
  useState,
} from 'react';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { z } from 'zod';

import { articleSchema } from '@/models/articleSchema';
import { createArticleAction } from '@/app/articleActions';
import { UrlsTypes } from '@/models/article';
import ArticleMarkupForm from '@/app/components/single-elements/articleHTMLForm';
import { urlsToArrayUtil } from '@/lib/utility-functions';
import NotificationsComponent from '@/app/components/single-elements/notificationsComponent';
import { customResolver } from '@/app/editor/components/resolvers/customResolver';
import { ArticleTitle } from '@/app/components/single-elements/ArticleTitle';

/**
 * CreateArticleForm is a React component that manages the creation of an article form.
 * It handles form state, validation, submission, and various utility functions for managing URLs.
 *
 * @returns {JSX.Element} Returns an ArticleMarkupForm component with all necessary props for creating an article.
 */
export default function CreateArticleForm({
  scrolltoTop,
}: {
  scrolltoTop: () => void;
}): JSX.Element {
  const [state, formAction, isPending] = useActionState(
    createArticleAction,
    null
  );
  const [notification, setNotification] = useState<boolean>(false);

  // declaring react hook form variables
  const {
    register,
    handleSubmit,
    watch,
    clearErrors,
    setValue,
    getValues,
    reset,
    formState: { errors },
  } = useForm<z.infer<typeof articleSchema>>({
    mode: 'onChange',
    reValidateMode: 'onChange',
    resolver: customResolver(articleSchema) as any,
    defaultValues: {
      title: '',
      introduction: '',
      main: '',
      mainAudioUrl: '',
      urlToMainIllustration: '',
      urls: '',
    },
  });

  // registering urls that need a special treatment
  // adding, removing and updating those fields are handled by the urlsToArrayUtil function
  register('urls');
  const urlsToArray = urlsToArrayUtil(getValues('urls'));

  // main function that submits data to the server action part
  const onSubmit = (data: z.infer<typeof articleSchema>) => {
    startTransition(() => {
      formAction(data);
    });
  };

  useEffect(() => {
    // utility that helps clear errors when addressed live by the user
    const subscription = watch((value, { name }) => {
      if (name) {
        clearErrors(name);
      }
    });

    // display of the final notification, on success or failure
    let notifTimeout: NodeJS.Timeout | undefined;
    if (state) {
      setNotification(true);
      scrolltoTop();
      notifTimeout = setTimeout(() => {
        reset();
        setNotification(false);
      }, 6000);
    }

    return () => {
      subscription.unsubscribe();
      if (notifTimeout) {
        clearTimeout(notifTimeout);
      }
    };
  }, [watch, clearErrors, state]);

  // template for urls, injected when a url field is added
  const initialUrls = {
    type: 'website' as UrlsTypes,
    url: '',
    credits: '',
  };

  // utilities for urls fields
  const addInputs = () => {
    const urls = urlsToArray;
    setValue('urls', JSON.stringify([...urls, initialUrls]));
  };
  const removeInputs = () => {
    if (urlsToArray.length > 1) {
      setValue('urls', JSON.stringify(urlsToArray.slice(0, -1)));
    }
  };
  const updateUrls = (
    newUrl: { type: UrlsTypes; url: string; credits?: string },
    index: number
  ) => {
    const newUrls = urlsToArray;
    newUrls[index] = newUrl;
    setValue('urls', JSON.stringify(newUrls));
  };

  return (
    <>
      <ArticleTitle
        text="CrÃ©er un article"
        level="h2"
        size="large"
        color="white"
        spacings="mt-6 mb-4"
      />
      {notification && (
        <NotificationsComponent
          state={state as { message: boolean; text: string }}
        />
      )}
      <ArticleMarkupForm
        handleSubmit={handleSubmit(onSubmit)}
        register={register}
        errors={errors}
        urlsToArray={urlsToArray}
        updateUrls={updateUrls}
        addInputs={addInputs}
        removeInputs={removeInputs}
        target="create"
        isPending={isPending}
      />
    </>
  );
}
