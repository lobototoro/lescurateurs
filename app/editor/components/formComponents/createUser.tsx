import { zodResolver } from '@hookform/resolvers/zod';
import { startTransition, useActionState, useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { z } from 'zod';

import { ArticleTitle } from '@/app/components/single-elements/ArticleTitle';
import UserPermissionsCheckboxes from '@/app/components/single-elements/userPermissions';
import {
  adminPermissions,
  contributorPermissions,
  UserRole,
  userRoles,
} from '@/models/user';
import { userSchema } from '@/models/userSchema';
import { createUserAction } from '@/app/userActions';
import { customResolver } from '@/app/editor/components/resolvers/customResolver';

export default function CreateUserForm({
  scrolltoTop,
}: {
  scrolltoTop: () => void;
}) {
  const [state, formAction, isPending] = useActionState(createUserAction, null);
  const [userRole, setUserRole] = useState<keyof typeof UserRole>(
    userRoles[1] as unknown as keyof typeof UserRole
  );
  const [notification, setNotification] = useState<boolean>(false);

  const {
    register,
    handleSubmit,
    setValue,
    reset,
    watch,
    clearErrors,
    formState: { errors },
  } = useForm<z.infer<typeof userSchema>>({
    mode: 'onChange',
    reValidateMode: 'onBlur',
    resolver: customResolver(userSchema) as any,
    defaultValues: {
      email: '',
      tiers_service_ident: '',
      role: 'contributor',
      created_at: new Date().toISOString(),
      last_connection_at: new Date().toISOString(),
      permissions: JSON.stringify(contributorPermissions),
    },
  });

  const onSubmit = (data: z.infer<typeof userSchema>) => {
    startTransition(() => {
      const formData = new FormData();
      formData.append('email', data.email);
      formData.append('tiers_service_ident', data.tiers_service_ident);
      formData.append('role', data.role);
      formData.append('permissions', data.permissions);
      formData.append('created_at', data.created_at);
      formData.append('last_connection_at', data.last_connection_at);
      formAction(formData);
    });
  };

  register('permissions', { required: true });

  useEffect(() => {
    const subscription = watch((value, { name }) => {
      if (name) {
        clearErrors(name);
      }
    });

    let notifTimeout: NodeJS.Timeout | undefined;
    if (state) {
      setNotification(true);
      scrolltoTop();
      reset();
      notifTimeout = setTimeout(() => {
        setNotification(false);
      }, 6000);
    }

    return () => {
      if (notifTimeout) {
        clearTimeout(notifTimeout);
      }

      subscription.unsubscribe();
    };
  }, [state, watch, clearErrors]);

  return (
    <section className="section">
      {notification && (
        <div
          className={`notification ${state && state.message ? 'is-success' : 'is-danger'}`}
        >
          <p className="content">
            {state && state.message ? 'Succès' : 'Erreur'}:{' '}
            {state && state.text}
            <br />
            <span>Cette notification se fermera d'elle-même</span>
          </p>
        </div>
      )}
      <ArticleTitle
        text="Créer un utilisateur"
        level="h2"
        color="white"
        size="large"
        spacings="mb-6"
      />
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className="fixed-grid">
          <div className="grid">
            <div className="cell mb-6">
              <div className="field">
                <label htmlFor="Email" className="mr-2">
                  Email:
                </label>
                <input
                  id="Email"
                  type="email"
                  className="input mt-4"
                  data-testid="email"
                  {...register('email')}
                />
                {errors.email && (
                  <p className="has-text-danger">{errors.email.message}</p>
                )}
              </div>
            </div>
            <div className="cell mb-6">
              <div className="field">
                <label htmlFor="tiersServiceIdent" className="mr-2 mb-4">
                  Identifiant Tiers Service:
                </label>
                <input
                  id="tiersServiceIdent"
                  type="text"
                  className="input mt-4"
                  data-testid="tiersServiceIdent"
                  {...register('tiers_service_ident', {
                    required: true,
                  })}
                />
                {errors.tiers_service_ident && (
                  <p className="has-text-danger">
                    {errors.tiers_service_ident.message}
                  </p>
                )}
              </div>
            </div>
            <div className="cell">
              <div className="field is-flex is-align-items-center">
                <label htmlFor="role" className="mr-4">
                  Rôle:
                </label>
                <div className="select">
                  <select
                    id="role"
                    data-testid="role"
                    {...register('role')}
                    onChange={(e) => {
                      setUserRole(e.target.value as keyof typeof UserRole);
                      if (e.target.value === 'admin') {
                        setValue(
                          'permissions',
                          JSON.stringify(adminPermissions)
                        );
                      }
                      if (e.target.value === 'contributor') {
                        setValue(
                          'permissions',
                          JSON.stringify(contributorPermissions)
                        );
                      }
                    }}
                    className="select"
                  >
                    {userRoles.map((role) => (
                      <option
                        key={role}
                        value={role}
                        className="text-sm has-text-white"
                      >
                        {role.charAt(0).toUpperCase() + role.slice(1)}
                      </option>
                    ))}
                  </select>
                </div>
                {errors.role && (
                  <p className="has-text-danger">{errors.role.message}</p>
                )}
              </div>
            </div>
            <div className="cell">
              {userRole && (
                <UserPermissionsCheckboxes role={[userRole as UserRole]} />
              )}
            </div>
          </div>
        </div>

        <button
          role="button"
          data-testid="final-submit"
          type="submit"
          className={
            isPending
              ? 'button is-primary is-size-6 has-text-white mt-5 is-loading'
              : 'button is-primary is-size-6 has-text-white mt-5'
          }
        >
          {isPending ? 'Chargement...' : "Créer l'utilisateur"}
        </button>
      </form>
    </section>
  );
}
