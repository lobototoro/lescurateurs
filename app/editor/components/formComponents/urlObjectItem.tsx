"use client";
import React from "react";

import { UrlsTypes } from "@/models/article";

export function UrlObjectItem({
  type,
  url,
  credits,
  urls,
  index,
  addUrls
}: {
  type?: UrlsTypes;
  url?: string;
  credits?: string;
  urls?: { type: UrlsTypes; url: string; credits?: string }[];
  index: number;
  addUrls?: (newUrl: { type: UrlsTypes; url: string; credits?: string }, index: number) => void;
}) {
  const [selectedValue, setSelectedValue] = React.useState<UrlsTypes>(type || UrlsTypes.WEBSITE);
  const [givenUrl, setGivenUrl] = React.useState<string>(url || '');
  const [givenCredits, setGivenCredits] = React.useState<string>(credits || '');

  const isinArray = (url: string) => {
    return urls?.some((item) => item.url === url && item.url !== '');
  };

  return (
    <div className="field is-grouped is-flex is-align-items-center">
      <div className="">
        <label className="is-hidden">Selectionnez un type</label>
        <div className="select">
          <select
            role="combobox"
            aria-label="Selectionnez un type"
            value={selectedValue || UrlsTypes.WEBSITE}
            onChange={(e) => setSelectedValue(e.target.value as UrlsTypes)}
          >
            <option value={UrlsTypes.WEBSITE}>website</option>
            <option value={UrlsTypes.VIDEOS}>videos</option>
            <option value={UrlsTypes.AUDIO}>audio</option>
            <option value={UrlsTypes.SOCIAL}>sociaux</option>
            <option value={UrlsTypes.IMAGE}>image</option>
          </select>
        </div>
      </div>
      <div className="ml-2">
        <label className="mr-2" role="label">url</label>
        <input
          role="textbox"
          type="url"
          name="url"
          value={givenUrl || ""}
          onChange={(e) => setGivenUrl(e.target.value)}
        />
      </div>
      <div className="ml-2">
        <label className="mr-2" role="label">cr√©dits</label>
        <input
          role="textbox"
          type="text"
          name="credits"
          value={givenCredits || ""}
          onChange={(e) => setGivenCredits(e.target.value)}
        />
      </div>
      <div
        className="button ml-2"
        onClick={() => addUrls && addUrls({
              type: selectedValue,
              url: givenUrl,
              credits: givenCredits
            },
            index
          )}
      >
        {
          (isinArray(givenUrl)) ? 'Modifier' : 'Ajouter'
        }
      </div>
    </div>
  );
}