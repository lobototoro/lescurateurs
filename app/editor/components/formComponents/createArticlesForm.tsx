"use client";
import React from "react";

import { AddUrlsObjects } from "./addUrlsObjects";

// import { UrlObjectIterator } from "./urlObjectIter";
import { UrlsTypes } from '@/models/article';
import { ArticleTitle } from "@/app/components/single-elements/ArticleTitle";

export default function CreatearticlesForm() {
  const [urls, setUrls] = React.useState<{ type: UrlsTypes; url: string; credits?: string }[]>([]);

  const initialUrls = {
      type: 'website' as UrlsTypes,
      url: '',
      credits: '',
  };
  
  const addInputs = () => {
    setUrls([...urls, initialUrls]);
  }
  const removeInputs = () => {
    if (urls.length > 1) {
      setUrls(urls.slice(0, -1));
    }
  }

  const updateUrls = (newUrl: { type: UrlsTypes; url: string; credits?: string }, index: number) => {
    const newUrls = [...urls];
    newUrls[index] = newUrl;
    setUrls(newUrls);
  };
  
  // console.log('in create article ', urls);
  
  return (
    <div className="mt-6">
      <div className="field">
      <label className="label">Titre</label>
      <div className="control">
        <input className="input" type="text" placeholder="titre" name="title" />
      </div>
    </div>

    <div className="field">
      <label className="label">introduction</label>
      <textarea className="textarea is-small" placeholder="introduction" name="introduction"></textarea>
    </div>

    <div className="field">
      <label className="label">Texte</label>
      <textarea className="textarea is-large" placeholder="Texte" name="main" rows={10} cols={40}></textarea>
    </div>

    <div className="field">
      <label className="label">lien audio principal</label>
      <div className="control">
        <input className="input" type="text" placeholder="lien audio" name="mainAudioUrl" />
      </div>
    </div>

    <div className="field">
      <label className="label is-inline-flex">lien vers l'illustration</label>
      <div className="control">
        <input className="input" type="text" placeholder="lien vers l'illustration" name="urlToMainIllustration" />
      </div>
    </div>
    <ArticleTitle
      text="Ajouter des liens multimÃ©dias"
      level="h4"
      size="medium"
      color="white"
      spacings="mt-5 mb-4"
    />
    {/* <AddUrlsObjects /> */}
    {/* <UrlObjectIterator
      addUrls={setUrls}
      type={'audio' as UrlsTypes}
      url="https://google.com"
      credits="google"
    /> */}
    <AddUrlsObjects
      urls={urls}
      updateUrls={updateUrls}
      addInputs={addInputs}
      removeInputs={removeInputs}
    />
  </div>);
}