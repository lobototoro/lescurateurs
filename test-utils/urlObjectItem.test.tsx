import { UrlObjectItem } from "@/app/editor/components/formComponents/urlObjectItem";
import { UrlsTypes } from "@/models/article";
import { cleanup, render } from "@testing-library/react";
import { afterEach, describe, expect, it } from "vitest";

describe("checking rendering and data integrety", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render correctly", () => {
    const { asFragment } = render(
    <UrlObjectItem
      type={UrlsTypes.WEBSITE}
      url=""
      credits=""
      urls={[]}
      index={0}
      addUrls={() => {}}
    />
  );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should return an object filled with the correct values when 'ajouter' is clicked", () => {
    const { getByText, queryByDisplayValue } = render(
    <UrlObjectItem
      type={UrlsTypes.WEBSITE}
      url="https://example.com"
      credits="pablo"
      urls={[]}
      index={0}
      addUrls={() => {}}
    />
  );

    expect(getByText("Ajouter")).toBeDefined();
    expect(queryByDisplayValue("https://example.com")).toBeTruthy();
    expect(queryByDisplayValue("pablo")).toBeTruthy();
  });
});