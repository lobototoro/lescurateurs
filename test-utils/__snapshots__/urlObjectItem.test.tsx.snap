// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checking rendering and data integrety > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="field is-grouped is-flex is-align-items-center"
  >
    <div
      class=""
    >
      <label
        class="is-hidden"
      >
        Selectionnez un type
      </label>
      <div
        class="select"
      >
        <select
          aria-label="Selectionnez un type"
          name="type-0"
          role="combobox"
        >
          <option
            value="website"
          >
            website
          </option>
          <option
            value="videos"
          >
            videos
          </option>
          <option
            value="audio"
          >
            audio
          </option>
          <option
            value="social"
          >
            sociaux
          </option>
          <option
            value="image"
          >
            image
          </option>
        </select>
      </div>
    </div>
    <div
      class="ml-2"
    >
      <label
        class="mr-2"
      >
        url
      </label>
      <input
        name="url-0"
        type="url"
        value=""
      />
    </div>
    <div
      class="ml-2"
    >
      <label
        class="mr-2"
        role="label"
      >
        cr√©dits
      </label>
      <input
        name="credits-0"
        type="text"
        value=""
      />
    </div>
    <div
      class="button ml-2"
    >
      Ajouter
    </div>
  </div>
</DocumentFragment>
`;
